CyberArk User Management

cyberark-bizdev opened the pull request:
SUMMARY

Added Module:

cyberark_user: details, add/update, delete, addtogroup for CyberArk Users

This module uses CyberArk Privileged Account Security Web Services
SDK

ISSUE TYPE

New Module Pull Request
COMPONENT NAME

lib/ansible/modules/identity/cyberark/cyberark_user.py

ANSIBLE VERSION

ansible 2.2.0.0
config file = /etc/ansible/ansible.cfg
configured module search path = Default w/o overrides

ADDITIONAL INFORMATION

enunez-cyberark committed:
Initial submit for CyberArk User management module

enunez-cyberark committed:
Updated Documentation section

enunez-cyberark committed:
Initial Submit

enunez-cyberark committed:
Changed "RETURN" documentation section.

enunez-cyberark committed:
Removed payload reference in userDetails and userDelete module

enunez-cyberark committed:
Pep8 and Yamllint errors by shippable.

enunez-cyberark committed:
Updated Documentation section

enunez-cyberark committed:
Another fix to Documentation section

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,515 @@
+#!/usr/bin/python
+#
+# This file is part of Ansible
```
There should be a copyright line in here.  It can be as simple as 
```
# Copyright 2017 Ansible Project
```
if you want.  For any legal questions, we would have to use the git log to find out the real copyright holders so a catchall like above is okay.

cyberark-bizdev replied:
Done

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,515 @@
+    import httplib
+except ImportError:
+    # Python 3
+    import http.client as httplib
```
note: standard python style is to make three groups of imports.  libraries from the standard library first, then imports from third party libraries, and finally imports from the project you're a part of.  imports should also be roughly alphabetical.  So the order of imports should look more like this:
``` python
import json
try:
    import httplib
except ImportError:
    # Python 3
    import http.client as httplib

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils._text import to_text
from ansible.module_utils.six.moves.urllib.error import HTTPError
from ansible.module_utils.urls import open_url
```

(httpclient, because it has the try except block is not included in the alphabetical sorting... see my comment specifically on http.client though.)

cyberark-bizdev replied:
Done

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,515 @@
+    import httplib
+except ImportError:
+    # Python 3
+    import http.client as httplib
```
There's an entry in six.moves for this.  Probably better to use:
``` python
from ansible.module_utils.six.moves import http_client

OR

from ansible.module_utils.six.moves import http_client as httplib
```

cyberark-bizdev replied:
Done

enunez-cyberark committed:
Updates based on community feedback.

Changed to support present and absent. removed details, addtogroup, and
update

enunez-cyberark committed:
Interpreter line and license updated

enunez-cyberark committed:
PEP8

enunez-cyberark committed:
Added __future__ import ...

cyberark-bizdev commented:
ready_for_review

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,461 @@
+                 "\n*** end_point=%s%s\n%s" % (api_base_url, end_point, to_text(unknown_exception))),
+            payload=payload,
+            headers=headers,
+            status_code=-1)
```
Anytime you handle an unknown error, it's good to include the traceback.  Do it like this:
``` python
import traceback
[....]

module.fail_json(
            msg=("Unknown error while performing userAddToGroup."
                 "\n*** end_point=%s%s\n%s" % (api_base_url, end_point, to_text(unknown_exception))),
            payload=payload,
            headers=headers,
            exception=traceback.format_exc(),
            status_code=-1)
```

abadger replied:
There's other places that unknown exceptions are handled.  You should modify those similar to this.

cyberark-bizdev replied:
done!

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,461 @@
+    sample: 200
+'''
+
+import json
```
Just style, there should be a blank line between stdlib imports and the next set of imports (in this case, imports from ansible)

cyberark-bizdev replied:
done

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,461 @@
+from ansible.module_utils.urls import open_url
+
+
+def userDetails(module):
```
Just style again.  We use this style for most identifiers (including function names):
```
def user_details(module):
```

For classes we use CapWord like:
``` python
class PasswordSession:
    pass
```

We don't use initial-lower camelCase anywhere.

cyberark-bizdev replied:
changed

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,461 @@
+        description:
+            - The name of the user who will be queried (for details), added, updated or deleted.
+    state:
+        default: details
```
The default should either be one of the choices or None if this is a tristate (ie: not being explicitly set by the user has a special meaning).

cyberark-bizdev replied:
fixed. This was a leftover from a previous update (make the module declarative)

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,461 @@
+        "username": {"required": True, "type": "str"},
+        "state": {"type": "str",
+                  "choices": ["details", "present", "update", "addtogroup", "absent"],
+                  "default": "details"},
```
The choices here don't seem to match with the documentation or the code.  The default doesn't match with the code.

cyberark-bizdev replied:
fixed. This was a leftover from a previous update (make the module declarative)

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,461 @@
+    change_password_on_the_next_logon: false
+    group_name: "GroupOfUsers"
+    state: present
+    cyberark_session: "{{ cyberark_session }}"
```
There should be an example of ```cyberark_session``` somewhere.  So that users know what to put in it.

cyberark-bizdev replied:
in the parameter definition, I mentioned to check module cyberark_authentication for example of cyberark_session

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,461 @@
+
+    if module.check_mode:
+        # return if case of check_mode
+        module.exit_json(change=False)
```
If we don't know whether a change would be made or not, we should always return changed=True.  It's better for a user to run their playbook unnecessarily than for a user to miss out on running a playbook to make changes because check_mode reported False.

abadger replied:
Note that there are other pieces of code that need this same fix.

abadger commented:
Okay, I think I've found everything.  Once those are fixed, feel free to ping me on IRC (abadger1999) to have another look.

abadger commented:
needs_revision

enunez-cyberark committed:
Updates based on community feedback (@abadger)

enunez-cyberark committed:
Fixed state choices and default

cyberark-bizdev commented:
ready_for_review

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_user.py:
```diff
@@ -0,0 +1,456 @@
+        required: True
+        description:
+            - Dictionary set by a CyberArk authentication containing the different values to perform actions on a logged-on CyberArk session,
+              please see cyberark_authentication module for an example of cyberark_session.
```
IIRC, you can use ```M(cyberark_authentication)``` to link to the cyberark_authentication module documentation

abadger replied:
yeah, M(cyberark_authentication) is how to link to another module.

cyberark-bizdev replied:
Done

abadger commented:
Okay, just hte note about linking to cyberark_authentication docs and also the other functions should be renamed from camelcase to underscore separated words like
```
def userAddToGroup(module):  => def user_add_to_group(module):
```

enunez-cyberark committed:
Changes requested (abadger1999)

- Function names from camelcase to underscore
- Link to doc of cyberark_authentication

abadger commented:
Merged to devel for 2.4.0 release.  Thanks!
